src/components/timeline/TrackScreenshots.js

```diff
@@ -300,6 +122,132 @@ class HoverPreview extends PureComponent<HoverPreviewProps> {
   }
 }
 
+type Props = {|
+  +threadIndex: ThreadIndex,
+  +windowId: string,
+|};
+
+export function TimelineTrackScreenshots({ threadIndex, windowId }: Props) {
+  // Selectors
+  const selectors = getThreadSelectors(threadIndex);
+  const { start: rangeStart, end: rangeEnd } = useSelector(getCommittedRange);
+  const isMakingPreviewSelection = useSelector(getIsMakingPreviewSelection);
+  const thread = useSelector(selectors.getRangeFilteredThread);
+  const screenshotsById = useSelector(
+    selectors.getRangeFilteredScreenshotsById
+  );
+  const trackHeight = useSelector(getScreenshotTrackHeight);
+  const screenshots = screenshotsById.get(windowId) || EMPTY_SCREENSHOTS_TRACK;
+
+  // Hooks
+  const dispatch = useDispatch<Dispatch>();
+  const {
+    ref,
+    dimensions: { width },
+  } = useDimensions();
+
+  // State
+  const [pageX, setPageX] = useState<CssPixels | null>(null);
+  const [offsetX, setOffsetX] = useState<CssPixels | null>(null);
+  const [containerTop, setContainerTop] = useState<CssPixels | null>(null);
+
+  function handleMouseLeave() {
+    setPageX(null);
+    setOffsetX(null);
+    setContainerTop(null);
+  }
+
+  function handleMouseMove(event: SyntheticMouseEvent<HTMLDivElement>) {
+    const { top, left } = event.currentTarget.getBoundingClientRect();
+    setPageX(event.pageX);
+    setOffsetX(event.pageX - left);
+    setContainerTop(top);
+  }
+
+  // This selects a screenshot when clicking on the screenshot strip. Note that
+  // we use the mouseup event so that isMakingPreviewSelection is still
+  // accurate.
+  function selectScreenshotOnClick(event: SyntheticMouseEvent<HTMLDivElement>) {
+    if (isMakingPreviewSelection) {
+      // Avoid reseting the selection if the user is currently selecting one.
+      return;
+    }
+
+    const { left } = event.currentTarget.getBoundingClientRect();
+    const offsetX = event.pageX - left;
+    const screenshotIndex = findScreenshotAtMouse(offsetX);
+    if (screenshotIndex === null) return;
+    const { start, end } = screenshots[screenshotIndex];
+    if (end === null) return;
+    dispatch(
+      updatePreviewSelection({
+        hasSelection: true,
+        isModifying: false,
+        selectionStart: start,
+        selectionEnd: end,
+      })
+    );
+  }
+
+  function findScreenshotAtMouse(offsetX: number): number | null {
+    const rangeLength = rangeEnd - rangeStart;
```

Comment:
This is capturing state in the function! Crazy. However, I'm not using a memoized version of this, since it's going into a div, not a PureComponent. I tried to decide about adding these outside of the function scope. I think their is a trade-off between extra wiring into functions, and mistakenly capturing stale values.

The second parameter of `useCallback` is an invalidation list, that will re-run the function when certain values have changed. I didn't have the need for it here.

Reply:
> This is capturing state in the function! 

But isn't the function redefined each time the state changes anyway (because it's rerendered)? So I don't think the state could be stale.
But BTW could that be a performance problem?